src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,475 @@
+  public void clean() throws ExecException, InterruptedException {
+    CleanRequest.Builder builder = CleanRequest.newBuilder();
+    builder.setOutputBaseId(outputBaseId);
+    blockingStub.clean(builder.build());
+  }
+
+  @Override
+  public ActionFileSystemType actionFileSystemType() {
+    return ActionFileSystemType.STAGE_REMOTE_FILES;
+  }
+
+  private class GrpcFileSystem extends DelegateFileSystem {
+    private final Path originalOutputPath;
+
+    public GrpcFileSystem(FileSystem sourceDelegate, PathFragment originalOutputPath) {
+      super(sourceDelegate);
+      this.originalOutputPath = getPath(originalOutputPath);
+    }
+
+    @Override
+    protected byte[] getFastDigest(Path path) throws IOException {
+      // Don't attempt to compute digests ourselves. Call into the
+      // remote output service to request the digest. The service may be
+      // able to return the digest from its bookkeeping, as opposed to
+      // reading the actual file contents.
+      BatchStatRequest.Builder builder = BatchStatRequest.newBuilder();
+      builder.setBuildId(currentBuildId.toString());
+      builder.setIncludeFileDigest(true);
+      builder.setFollowSymlinks(true);
+      PathFragment relativePath;
+      try {
+        relativePath = path.relativeTo(originalOutputPath);
+      } catch (IllegalArgumentException e) {
+        // Path is outside the output path. Send it to the regular file system.
+        return super.getFastDigest(path);
+      }
+      builder.addPaths(relativePath.toString());
+
+      BatchStatResponse responses = blockingStub.batchStat(builder.build());
+      StatResponse response = responses.getResponses(0);
+      if (!response.hasFileStatus()) {
+        // Output service reports that the file does not exist.
+        throw new FileNotFoundException();
+      }
+      FileStatus fileStatus = response.getFileStatus();
+      if (fileStatus.hasFile()) {
+        FileStatus.File regularFileStatus = fileStatus.getFile();
+        return DigestUtil.toBinaryDigest(regularFileStatus.getDigest());
+      }
+      if (fileStatus.hasExternal()) {
+        // Path is a symbolic link that points to a location outside of
+        // the output path.  The output service is unable to give a
+        // digest. We must compute it ourselves.
+        return super.getFastDigest(
+            originalOutputPath.getRelative(fileStatus.getExternal().getNextPath()));
+      }
+      throw new IOException(format("Remote output service did not return the status of file %s", path));
+    }
+  }
+
+  @Override
+  @Nullable
+  public FileSystem createActionFileSystem(
+      FileSystem sourceDelegate,
+      PathFragment execRootFragment,
+      String relativeOutputPath,
+      ImmutableList<Root> sourceRoots,
+      ActionInputMap inputArtifactData,
+      Iterable<Artifact> outputArtifacts,
+      boolean trackFailedRemoteReads) {
+    return new GrpcFileSystem(sourceDelegate, execRootFragment.getRelative(relativeOutputPath));
```

Comment:
This only takes care of the case where files are accessed as part of actions. Another case we need to tackle is BES uploads. Now I tried to add these functions:
```
diff --git a/src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java b/src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java
index 479e3b5a05..a4c67d45d0 100644
--- a/src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java
+++ b/src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java
@@ -22,6 +22,7 @@ import com.google.common.util.concurrent.MoreExecutors;
 import com.google.devtools.build.lib.actions.Action;
 import com.google.devtools.build.lib.actions.ActionInputMap;
 import com.google.devtools.build.lib.actions.Artifact;
+import com.google.devtools.build.lib.actions.ArtifactPathResolver;
 import com.google.devtools.build.lib.actions.BuildFailedException;
 import com.google.devtools.build.lib.actions.cache.MetadataHandler;
 import com.google.devtools.build.lib.actions.EnvironmentalExecException;
@@ -513,4 +514,23 @@ public class GrpcRemoteOutputService implements OutputService, ActionResultDownl
         (result) -> null,
         MoreExecutors.directExecutor());
   }
+
+  @Override
+  public boolean supportsPathResolverForArtifactValues() {
+    return true;
+  }
+
+  @Override
+  public ArtifactPathResolver createPathResolverForArtifactValues(
+      PathFragment execRoot,
+      String relativeOutputPath,
+      FileSystem fileSystem,
+      ImmutableList<Root> pathEntries,
+      ActionInputMap actionInputMap,
+      Map<Artifact, Collection<Artifact>> expandedArtifacts,
+      Map<Artifact, ImmutableList<FilesetOutputSymlink>> filesets) {
+    FileSystem remoteFileSystem =
+        new GrpcFileSystem(fileSystem, execRoot.getRelative(relativeOutputPath));
+    return ArtifactPathResolver.createPathResolver(remoteFileSystem, fileSystem.getPath(execRoot));
+  }
 }
```

But I've observed that this breaks in an interesting way: BES uploads may continue to run after OutputService.finalizeBuild() is called, meaning we can no longer call BatchStat().

Question: Is there some way to detect that the PathResolver returned by createPathResolverForArtifactValues() is released, or is there simply no such guarantee?

Reply:
It looks like I've managed to get integration with Build Event Streams working. In addition to afterCommand(), there is this commandComplete() method that we can use to safely call the FinalizeBuild() RPC.